python/shotgun_data/sgdata.py

```diff
@@ -272,53 +273,32 @@ def request_thumbnail(self, url, entity_type, entity_id, field):
     def _get_thumbnail_path(url, bundle):
         """
         Returns the location on disk suitable for a thumbnail given its url.
+        
+        :param bundle: App, Engine or Framework instance
+        :param url: Path to a thumbnail
+        :returns: Path as a string.
         """
-
+        # hash the path portion of the thumbnail url
         url_obj = urlparse.urlparse(url)
-        url_path = url_obj.path
-        path_chunks = url_path.split("/")
-
-        CHUNK_LEN = 16
-
-        # post process the path
-        # old (pre-S3) style result:
-        # path_chunks: [ "", "thumbs", "1", "2", "2.jpg"]
-
-        # s3 result, form 1:
-        # path_chunks: [u'',
-        #               u'9902b5f5f336fae2fb248e8a8748fcd9aedd822e',
-        #               u'be4236b8f198ae84df2366920e7ee327cc0a567e',
-        #               u'render_0400_t.jpg']
-
-        # s3 result, form 2:
-        # path_chunks: [u'', u'thumbnail', u'api_image', u'150']
-
-        def _to_chunks(s):
-            #split the string 'abcdefghxx' into ['abcdefgh', 'xx']
-            chunks = []
-            for start in range(0, len(s), CHUNK_LEN):
-                chunks.append( s[start:start+CHUNK_LEN] )
-            return chunks
-
-        new_chunks = []
-        for folder in path_chunks[:-1]: # skip the file name
-            if folder == "":
-                continue
-            if len(folder) > CHUNK_LEN:
-                # long url path segment like 9902b5f5f336fae2fb248e8a8748fcd9aedd822e
-                # split it into chunks for 4
-                new_chunks.extend( _to_chunks(folder) )
-            else:
-                new_chunks.append(folder)
+        url_hash = hashlib.md5()
+        url_hash.update(str(url_obj.path))
+        hash_str = url_hash.hexdigest()
+
+        # Now turn this hash into a tree structure. For a discussion about sensible
+        # sharding methodology, see 
+        # http://stackoverflow.com/questions/13841931/using-guids-as-folder-names-splitting-up
+        #
+        # From the hash, generate paths on the form C1C2/C3C4/full_hash.jpeg
+        # for a million evenly distributed items, this means ~15 items per folder
+        first_folder = str(hash_str[0:2])
+        second_folder = str(hash_str[2:4])
+        file_name = "%s.jpeg" % hash_str
```

Comment:
If you feel that it's worth making strings smaller at https://github.com/shotgunsoftware/tk-framework-shotgunutils/pull/12/files#diff-018892ec714588513fa837ee353c2f40R394, then maybe it would be wise to truncate the first 4 characters of the file name since they are already present in the folder path.